### DISCUSSION TITLE
About systemcall

### DISCUSSION THREAD

**Discussion by skandbug** [aquasecurity/tracee#2945] [Questions and Help]
Tracee is a great tool. But I don't know how to get the system calls that the program runs through her.
e.g.,:
ls command's: execve,brk,arch_prctl,mmap....

**Comment by geyslan** [comment]
> Tracee is a great tool.

@skandbug Thanks for loving tracee.

> But I don't know how to get the system calls that the program runs through her. e.g.,: ls command's: execve,brk,arch_prctl,mmap....

Please, try like this:

```
sudo ./dist/tracee -f comm=ls -f e=brk,arch_prctl,mmap
```

And running ls in other terminal tracee will output this:

```
TIME             UID    COMM             PID     TID     RET              EVENT                     ARGS
10:15:38:238468  1000   ls               476785  476785  94688011005952   brk                       addr: 0x0
10:15:38:238577  1000   ls               476785  476785  -22              arch_prctl                option: 12289, addr: 140721550348768
10:15:38:238664  1000   ls               476785  476785  140329938677760  mmap                      addr: 0x0, length: 171339, prot: PROT_READ, flags: 2, fd: 3, off: 0
10:15:38:238723  1000   ls               476785  476785  140329938669568  mmap                      addr: 0x0, length: 8192, prot: PROT_READ|PROT_WRITE, flags: 34, fd: -1, off: 0
10:15:38:238743  1000   ls               476785  476785  140329938620416  mmap                      addr: 0x0, length: 45128, prot: PROT_READ, flags: 2050, fd: 3, off: 0
10:15:38:238756  1000   ls               476785  476785  140329938632704  mmap                      addr: 0x7fa11c24c000, length: 20480, prot: PROT_READ|PROT_EXEC, flags: 2066, fd: 3, off: 12288
10:15:38:238792  1000   ls               476785  476785  140329938653184  mmap                      addr: 0x7fa11c251000, length: 8192, prot: PROT_READ, flags: 2066, fd: 3, off: 32768
10:15:38:238815  1000   ls               476785  476785  140329938661376  mmap                      addr: 0x7fa11c253000, length: 8192, prot: PROT_READ|PROT_WRITE, flags: 2066, fd: 3, off: 36864
10:15:38:238875  1000   ls               476785  476785  140329936625664  mmap                      addr: 0x0, length: 1994096, prot: PROT_READ, flags: 2050, fd: 3, off: 0
10:15:38:238890  1000   ls               476785  476785  140329936764928  mmap                      addr: 0x7fa11c084000, length: 1417216, prot: PROT_READ|PROT_EXEC, flags: 2066, fd: 3, off: 139264
10:15:38:238921  1000   ls               476785  476785  140329938182144  mmap                      addr: 0x7fa11c1de000, length: 360448, prot: PROT_READ, flags: 2066, fd: 3, off: 1556480
10:15:38:238943  1000   ls               476785  476785  140329938542592  mmap                      addr: 0x7fa11c236000, length: 24576, prot: PROT_READ|PROT_WRITE, flags: 2066, fd: 3, off: 1916928
10:15:38:238973  1000   ls               476785  476785  140329938567168  mmap                      addr: 0x7fa11c23c000, length: 52592, prot: PROT_READ|PROT_WRITE, flags: 50, fd: -1, off: 0
10:15:38:239033  1000   ls               476785  476785  140329936613376  mmap                      addr: 0x0, length: 12288, prot: PROT_READ|PROT_WRITE, flags: 34, fd: -1, off: 0
10:15:38:239051  1000   ls               476785  476785  0                arch_prctl                option: 4098, addr: 140329936615232
10:15:38:239482  1000   ls               476785  476785  94688011005952   brk                       addr: 0x0
10:15:38:239492  1000   ls               476785  476785  94688011141120   brk                       addr: 0x561e4544c000
10:15:38:239535  1000   ls               476785  476785  140329932029952  mmap                      addr: 0x0, length: 3061808, prot: PROT_READ, flags: 2, fd: 3, off: 0
```

For more see: https://aquasecurity.github.io/tracee/dev/docs/filters/filtering/

Closing as not an issue, but feel free to ask.

**Comment by skandbug** [comment]
However, if you trace an executable program, the system calls are not fixed.
How do we get the sequence of system calls for an application?

**Comment by yanivagman** [comment]
@skandbug you can use the "syscalls" set to get what you want, simply add the following filter:
`-f set=syscalls`

**Reply by skandbug** [reply]
Thanks!

**Reply by geyslan** [reply]
Considering all sorted (ordered) syscalls from ls binary:

`sudo ./dist/tracee -f binary=/usr/bin/ls -f set=syscalls -o option:sort-events`

**Reply by skandbug** [reply]
Thank you for the sorting scheme.
Is the default in the order in which the programs are executed?

**Reply by skandbug** [reply]
I found in my tests that tracee does not trace the system call "execve". I used "-f comm=ls -f e=execve"

**Reply by skandbug** [reply]
Is tracee designed to exclude the first system call "execve" when the program is executed by default?

**Reply by geyslan** [reply]
`ls` doesn't call execve, but does the shell which calls ls.

```
vagrant@ubuntu-jammy:/vagrant$ sudo ./dist/tracee -f comm=bash -f e=execve -f execve.args.pathname=*ls
TIME             UID    COMM             PID     TID     RET              EVENT                     ARGS
15:49:00:636723  1000   bash             106069  106069  0                execve                    pathname: /usr/bin/ls, argv: [ls --color=auto], envp: <nil>
```

**Reply by skandbug** [reply]
Do you get execve with this parameter?
-f comm=ls -f e=execve

**Reply by skandbug** [reply]
![image](https://user-images.githubusercontent.com/52226009/229172145-fc82f76d-595e-43bb-845e-890049e95c72.png)

here are my results

**Reply by geyslan** [reply]
`-f comm=ls -f e=execve` works as AND filters, instructing tracee to filter only execve calls from ls command. As I [mentioned](https://github.com/aquasecurity/tracee/discussions/2945#discussioncomment-5490941) ls doesn't call execve.

**Reply by skandbug** [reply]
I use` strace ls `and `-f comm=ls -f set=syscalls`
The difference between the two ways of getting system calls is that
strace captures the first syscall execve, while Tracee defaults to this one. All other sequences are the same.

**Reply by skandbug** [reply]
Again, please advise:
If I want to trace the system calls of an executable program, how can I get a complete and accurate system call (including the time)?

**Reply by yanivagman** [reply]
There shouldn't be any difference between strace and the output of tracee when using `-f set=syscalls` except the execve syscall shown by strace. As @geyslan wrote, this execve call isn't performed by the ls binary but by strace itself as it is executing the ls command. Tracee, on the other hand, doesn't perform the execve call but the shell from where you executed the command. Since the shell isn't traced as you added the `comm=ls` filter to tracee, you won't see this execve call. If you want to see it, simply add bash (or any other shell you are using) to the comm filter as in `-f comm=ls,bash -f set=syscalls`

**Reply by skandbug** [reply]
Thank you, and thank you to your team. Not only did you provide us with such a great tool, but you were able to answer our questions with dedication.

**Reply by skandbug** [reply]
![image](https://user-images.githubusercontent.com/52226009/229195687-b7679627-b617-49b2-b8df-dc10b69849e1.png)
This is the data format with json input. I wonder if Tracee supports output data cropping? As in the form of output in the terminal.
![image](https://user-images.githubusercontent.com/52226009/229195966-8205d326-a8f7-424e-89c5-ea42e252ec2a.png)

**Reply by yanivagman** [reply]
Nope. Currently json output gives all the context and data. In the future we plan to add custom output, where the user will be able to specify the fields of interest

**Reply by geyslan** [reply]
> ![image](https://user-images.githubusercontent.com/52226009/229195687-b7679627-b617-49b2-b8df-dc10b69849e1.png) This is the data format with json input. I wonder if Tracee supports output data cropping? As in the form of output in the terminal. ![image](https://user-images.githubusercontent.com/52226009/229195966-8205d326-a8f7-424e-89c5-ea42e252ec2a.png)

Perhaps this will help you: https://aquasecurity.github.io/tracee/dev/docs/outputs/output-formats/

![image](https://user-images.githubusercontent.com/3372117/229200043-2f200ccc-6dda-4319-b98a-4f63fb5d02ae.png)

**Reply by skandbug** [reply]
Thanks

**Reply by skandbug** [reply]
Tracee is a very good tool to use. But I don't know how well it tracks applications in docker containers or lxc containers.

**Reply by skandbug** [reply]
Such as tracking programs running in docker containers (e.g, browsers), Android apps running in lcx containers.

**Reply by skandbug** [reply]
https://aquasecurity.github.io/tracee/v0.14/classic-docs/docs/integrating/container-engines/

See the support for containers here.
But I don't know how the lxc container is supported, please tell me.

**Reply by rafaeldtinoco** [reply]
lxc/lxd containers aren't supported (just as regular OS processes) for now.

**Reply by skandbug** [reply]
Thanks!

**Reply by skandbug** [reply]
Such as: `strace ls -l | grep *.log`
When strace initiates a trace, it will execute the target command `ls -l | grep *.log`.
In the powerful tracee, do we provide a similar function.
